File: platform/cli-go/internal/commands/runner.go
package commands

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"syscall"

	"github.com/elloloop/devpod/platform/cli-go/internal/api"
	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/spf13/cobra"
)

func pidDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "share", "devpod")
}

func pidFile() string {
	return filepath.Join(pidDir(), "runner.pid")
}

func ensurePidDir() {
	_ = os.MkdirAll(pidDir(), 0o755)
}

func readPid() int {
	data, err := os.ReadFile(pidFile())
	if err != nil {
		return 0
	}
	pid, err := strconv.Atoi(strings.TrimSpace(string(data)))
	if err != nil {
		return 0
	}

	// Check if process is alive
	process, err := os.FindProcess(pid)
	if err != nil {
		_ = os.Remove(pidFile())
		return 0
	}

	// On Unix, FindProcess always succeeds. Check with signal 0.
	if runtime.GOOS != "windows" {
		if err := process.Signal(syscall.Signal(0)); err != nil {
			_ = os.Remove(pidFile())
			return 0
		}
	}

	return pid
}

func newRunnerCmd() *cobra.Command {
	runnerCmd := &cobra.Command{
		Use:   "runner",
		Short: "Manage the local runner",
	}

	// runner start
	var workspacePath string
	var port string

	startCmd := &cobra.Command{
		Use:   "start",
		Short: "Start the runner",
		RunE: func(cmd *cobra.Command, args []string) error {
			// Check if already running
			existingPid := readPid()
			if existingPid != 0 {
				if api.Ping() {
					fmt.Printf("Runner is already running %s\n", format.DimText(fmt.Sprintf("(PID %d)", existingPid)))
					return nil
				}
			}

			fmt.Printf("Starting runner on port %s for %s...\n", port, workspacePath)

			// Find the runner entry point relative to CLI location
			execPath, _ := os.Executable()
			cliDir := filepath.Dir(filepath.Dir(execPath))
			platformDir := filepath.Dir(cliDir)
			runnerDir := filepath.Join(platformDir, "runner")
			runnerEntry := filepath.Join(runnerDir, "src", "index.ts")

			var runnerCmdName string
			var runnerArgs []string

			if _, err := os.Stat(runnerEntry); err == nil {
				runnerCmdName = "npx"
				runnerArgs = []string{"tsx", runnerEntry}
			} else {
				runnerCmdName = "devpod-runner"
				runnerArgs = []string{}
			}

			child := exec.Command(runnerCmdName, runnerArgs...)
			child.Env = append(os.Environ(),
				"PORT="+port,
				"WORKSPACE="+workspacePath,
			)
			child.Dir = workspacePath
			child.Stdout = nil
			child.Stderr = nil
			child.Stdin = nil

			if err := child.Start(); err != nil {
				return fmt.Errorf("%s", format.ErrorMsg("Failed to start runner"))
			}

			if child.Process != nil {
				ensurePidDir()
				_ = os.WriteFile(pidFile(), []byte(strconv.Itoa(child.Process.Pid)), 0o644)
				// Detach the child
				_ = child.Process.Release()

				fmt.Printf("%s Runner started %s\n", format.SuccessMsg("\u2713"),
					format.DimText(fmt.Sprintf("(PID %d)", child.Process.Pid)))
				fmt.Println(format.DimText(fmt.Sprintf("  http://localhost:%s", port)))
			} else {
				return fmt.Errorf("%s", format.ErrorMsg("Failed to start runner"))
			}

			return nil
		},
	}

	cwd, _ := os.Getwd()
	startCmd.Flags().StringVarP(&workspacePath, "workspace", "w", cwd, "Workspace path")
	startCmd.Flags().StringVarP(&port, "port", "p", "4800", "Port number")

	// runner stop
	stopCmd := &cobra.Command{
		Use:   "stop",
		Short: "Stop the runner",
		RunE: func(cmd *cobra.Command, args []string) error {
			pid := readPid()
			if pid == 0 {
				fmt.Println(format.DimText("Runner is not running (no PID file found)"))
				return nil
			}

			process, err := os.FindProcess(pid)
			if err != nil {
				fmt.Println(format.DimText("Runner process not found, cleaning up PID file"))
				_ = os.Remove(pidFile())
				return nil
			}

			if err := process.Signal(syscall.SIGTERM); err != nil {
				fmt.Println(format.DimText("Runner process not found, cleaning up PID file"))
				_ = os.Remove(pidFile())
				return nil
			}

			_ = os.Remove(pidFile())
			fmt.Printf("%s Runner stopped %s\n", format.SuccessMsg("\u2713"),
				format.DimText(fmt.Sprintf("(PID %d)", pid)))
			return nil
		},
	}

	// runner status
	statusCmd := &cobra.Command{
		Use:   "status",
		Short: "Check runner status",
		RunE: func(cmd *cobra.Command, args []string) error {
			isUp := api.Ping()
			pid := readPid()

			if isUp {
				pidStr := ""
				if pid != 0 {
					pidStr = format.DimText(fmt.Sprintf(" (PID %d)", pid))
				}
				fmt.Printf("%s Runner is running%s\n", format.SuccessMsg("\u2713"), pidStr)
			} else {
				fmt.Printf("%s Runner is not running\n", format.ErrorMsg("\u2717"))
				fmt.Println(format.DimText("  Start with: devpod runner start"))
			}
			return nil
		},
	}

	runnerCmd.AddCommand(startCmd)
	runnerCmd.AddCommand(stopCmd)
	runnerCmd.AddCommand(statusCmd)

	return runnerCmd
}

File: platform/cli/src/commands/runner.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { spawn } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';
import * as api from '../api';
import { errorMessage } from '../format';

const PID_DIR = path.join(os.homedir(), '.local', 'share', 'devpod');
const PID_FILE = path.join(PID_DIR, 'runner.pid');

function ensurePidDir(): void {
  if (!fs.existsSync(PID_DIR)) {
    fs.mkdirSync(PID_DIR, { recursive: true });
  }
}

function readPid(): number | null {
  try {
    const content = fs.readFileSync(PID_FILE, 'utf-8').trim();
    const pid = parseInt(content, 10);
    if (isNaN(pid)) return null;

    // Check if process is alive
    try {
      process.kill(pid, 0);
      return pid;
    } catch {
      // Process not running, clean up stale PID file
      fs.unlinkSync(PID_FILE);
      return null;
    }
  } catch {
    return null;
  }
}

export function registerRunnerCommand(program: Command): void {
  const runner = program
    .command('runner')
    .description('Manage the local runner');

  // runner start
  runner
    .command('start')
    .description('Start the runner')
    .option('-w, --workspace <path>', 'Workspace path', process.cwd())
    .option('-p, --port <port>', 'Port number', '4800')
    .action(async (opts: { workspace: string; port: string }) => {
      try {
        // Check if already running
        const existingPid = readPid();
        if (existingPid) {
          const isUp = await api.ping();
          if (isUp) {
            console.log(
              `Runner is already running ${chalk.dim(`(PID ${existingPid})`)}`,
            );
            return;
          }
        }

        console.log(
          `Starting runner on port ${chalk.cyan(opts.port)} for ${chalk.cyan(opts.workspace)}...`,
        );

        // Find the runner entry point relative to CLI location
        const cliDir = path.resolve(__dirname, '..', '..');
        const platformDir = path.resolve(cliDir, '..');
        const runnerDir = path.join(platformDir, 'runner');
        const runnerEntry = path.join(runnerDir, 'src', 'index.ts');

        let runnerCmd: string;
        let runnerArgs: string[];

        if (fs.existsSync(runnerEntry)) {
          runnerCmd = 'npx';
          runnerArgs = ['tsx', runnerEntry];
        } else {
          // Fallback: assume runner is on PATH
          runnerCmd = 'devpod-runner';
          runnerArgs = [];
        }

        const child = spawn(runnerCmd, runnerArgs, {
          detached: true,
          stdio: 'ignore',
          env: {
            ...process.env,
            PORT: opts.port,
            WORKSPACE: opts.workspace,
          },
          cwd: opts.workspace,
        });

        if (child.pid) {
          ensurePidDir();
          fs.writeFileSync(PID_FILE, String(child.pid));
          child.unref();

          console.log(
            `${chalk.green('\u2713')} Runner started ${chalk.dim(`(PID ${child.pid})`)}`,
          );
          console.log(
            chalk.dim(`  http://localhost:${opts.port}`),
          );
        } else {
          console.error(chalk.red('Failed to start runner'));
          process.exit(1);
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // runner stop
  runner
    .command('stop')
    .description('Stop the runner')
    .action(async () => {
      try {
        const pid = readPid();
        if (!pid) {
          console.log(chalk.dim('Runner is not running (no PID file found)'));
          return;
        }

        try {
          process.kill(pid, 'SIGTERM');
          try {
            fs.unlinkSync(PID_FILE);
          } catch {
            // ignore
          }
          console.log(
            `${chalk.green('\u2713')} Runner stopped ${chalk.dim(`(PID ${pid})`)}`,
          );
        } catch {
          console.log(chalk.dim('Runner process not found, cleaning up PID file'));
          try {
            fs.unlinkSync(PID_FILE);
          } catch {
            // ignore
          }
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  // runner status
  runner
    .command('status')
    .description('Check runner status')
    .action(async () => {
      try {
        const runnerUrl = api.getRunnerUrl();
        const isUp = await api.ping();
        const pid = readPid();

        if (isUp) {
          const pidStr = pid ? chalk.dim(` (PID ${pid})`) : '';
          console.log(
            `${chalk.green('\u2713')} Runner is running at ${chalk.cyan(runnerUrl)}${pidStr}`,
          );
        } else {
          console.log(
            `${chalk.red('\u2717')} Runner is not running at ${chalk.dim(runnerUrl)}`,
          );
          console.log(
            chalk.dim('  Start with: devpod runner start'),
          );
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
